### PR TITLE
[v6] Add typescript definition for react-router@6

### PR DESCRIPTION
**Pull request by SevenOutman**
### Why?
I can't really use react-router@6 in my typescript project cuz there's no typescript definition for it.

### Why not DefinitelyTyped?
I saw in #6885 that typescript definition is on the roadmap.

### CODE REVIEW

**Comment by timdorr** [comment]
Also related to #7152.

File: packages/react-router/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  children?: RouteConfigObject[];
+}
+
+export interface MemoryRouterProps {
+  initialEntries?: string[];
+  initialIndex?: number;
+  timeout?: number;
+}
+
+export class MemoryRouter extends React.Componet<MemoryRouterProps, any> {
```
Comment (janovekj) [inline packages/react-router/index.d.ts]: A small typo :) `React.Componet` should be `React.Component`

**Comment by stunaz** [comment]
can this be merged? so TS users can test the v6 and also the definition at the same time?

**Comment by Pajn** [comment]
The types for react-router-dom does not export the props type for components. Please do that to not have to recreate these types when wrapping the components.

**Comment by CarsonF** [comment]
Hmm I think the v5 types in DefinitelyTyped would be a better starting point. They are more complete than these.

**Comment by SevenOutman** [comment]
> Hmm I think the v5 types in DefinitelyTyped would be a better starting point. They are more complete than these.

v6 and v5 have greatly different APIs and implementations. What’s “complete” in v5 definitions might not be even included in v6 logic. 

This edition of definitions are out source code of v6 implementation.

**Comment by SevenOutman** [comment]
> The types for react-router-dom does not export the props type for components. Please do that to not have to recreate these types when wrapping the components.

Added in e737bd3.

**Comment by ashubham** [comment]
Should we merge this ? I want to use v6 in my project. But being blocked due to Typescript definitions. cc @janovekj

File: packages/react-router/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+declare function navigate(delta: number): void;
+declare function navigate(to: History.LocationDescriptor, options?: NavigateOptions): void;
+
+/**
+ * Returns an imperative method for changing the location. Used by <Link>s, but
+ * may also be used by other elements to change the location.
+ */
+export function useNavigate(): typeof navigate;
```
Comment (gitpash) [inline packages/react-router/index.d.ts:130]: this is wip, just want to suggest some 
`type navigateFn<T> = (to: string | number, options?: NavigateOptions<T>) => void
export function useNavigate<T = Record<string, any>> (to: string | number, options?: Omit<NavigateProps<T>, 'to'>): navigateFn<T>;`

Comment (SevenOutman) [reply packages/react-router/index.d.ts:130]: The problem with `type navigateFn` you suggest is that it indicates that `navigate` has a call signature `navigate(to: number, options?)` which is actually not correct, according to the implementation. That's why I prefer `declare function`.

Adding generics support is a good idea. 😄

**Comment by mjackson** [comment]
Thank you for the work here, @SevenOutman. I'm sorry I didn't see this PR until today. I've been largely ignoring PRs to the dev branch until things settled a bit more. But I think the types I came up with in #7292 are actually pretty close to what you've done here, plus they are auto-generated from the source instead of being written by hand. If you wouldn't mind giving me some feedback there, I'd appreciate it very much.

Thanks!

**Pull request closed without merging**